### PR TITLE
Initial iteration: ModuleMinifierPlugin

### PR DESCRIPTION
**Pull request by dmichon-msft**
Adds a new package `@rushstack/module-minifier-plugin`. This package contains `ModuleMinifierPlugin`, which as a replacement for `terser-webpack-plugin` in the `optimization.minimizer` field of webpack config.

The general idea is to make the minification process more efficient through an alteration of the way that webpack communicates with Terser.

## How terser-webpack-plugin works
After webpack has rendered each chunk, the JS asset(s) corresponding to each chunk are handed over to Terser, then the minified result is returned to webpack and emitted.

## How @rushstack/module-minifier-plugin works
When webpack is performing final module rendering for the chunks, each module is extracted, hashed, and sent to Terser. When each chunk is being rendered, the object/array containing the modules for the chunk is replaced with an identifier `__WEBPACK_CHUNK_MODULES__` and the resulting JS asset (which contains only the chunk boilerplate, possibly the webpack runtime) is also sent to Terser. After all the modules and chunks have been processed, the plugin then replaces the `__WEBPACK_CHUNK_MODULES__` identifier with an array/object mapping ids to the minified modules.

## How to use (node 10 or 12)
```
// webpack.config.js
import { ModuleMinifierPlugin } from '@rushstack/module-minifier-plugin';
// This is not part of the main index import to prevent `worker_threads` from causing a compiler error in node 8
import { WorkerPoolMinifier } from '@rushstack/module-minifier-plugin/lib/WorkerPoolMinifier';

// In your webpack config object:
optimization: {
  minimizer: [new ModuleMinifierPlugin({ minifier: new WorkerPoolMinifier() })]
}
```

## How to use (node 8, or just for smaller scenarios where the thread overhead might be higher than the savings of parallelization)
```
// webpack.config.js
import { ModuleMinifierPlugin, SynchronousMinifier } form '@rushstack/module-minifier-plugin';

// In your webpack config object:
optimization: {
  minimizer: [new ModuleMinifierPlugin({ minifier: new SynchronousMinifier() })]
}
```

See @rushstack/module-minifier-plugin/readme.md for details on usage.

Current feature status:
- [x] Minifies code
- [x] Supports simple single-threaded operation mode
- [x] Compatible with html-webpack-plugin
- [x] Verified multi-threaded operation works (tested using OUFR)
- [x] Compatible with source maps (breakpoints in source seem to line up, and all source is visible in the debugger)
- [x] `usePortableModules` flag works as expected
- [x] Source map support now controlled by `sourceMap: true` flag for compatibility with other devtools than `source-map` and `inline-source-map`

### CODE REVIEW

**Comment by octogonz** [comment]
FYI we recently reformatted the master branch using Prettier, which creates merge conflicts for existing PRs. We'll try to fix up the existing PRs for people, but if you want to do it yourself, here's the steps:
 
1. Run "npm install -g prettier@2.0.5" to install Prettier.
2. Merge the "release/master-before-prettier" branch into your PR branch. This is the state of the world before Prettier was applied.
3. Merge the "release/master-after-prettier" branch into your PR branch. This will create a merge conflict with the Prettier changes, and nothing else.
4. Resolve all the merge conflicts by discarding "origin", and overwriting with your branch's changes.
5. BEFORE committing the merge, run "prettier . --write" from the repo root to reapply the formatting (that you discarded in step 4).
6. Make sure to restage everything.
7. Commit the merge
8. Now you should be able to merge from the real "master" without any (Prettier-related) merge conflicts

**Comment by dmichon-msft** [comment]
Looks like I need to fix some bad interactions between prettier and `//eslint-disable-line`

File: webpack/module-minifier-plugin/src/ModuleMinifierPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+              assets,
+              modules
+            },
+            compilation
+          );
+        }
+      );
+
+      for (const template of [compilation.chunkTemplate, compilation.mainTemplate]) {
+        // @ts-ignore Incompatible type definitions. Suffice to say, this hook exists.
```
Comment (iclanton) [inline webpack/module-minifier-plugin/src/ModuleMinifierPlugin.ts:410]: Can you just create another extended type here?

File: webpack/module-minifier-plugin/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./node_modules/@microsoft/rush-stack-compiler-3.5/includes/tsconfig-node.json",
+  "compilerOptions": {
+    "strictNullChecks": true,
```
Comment (iclanton) [inline webpack/module-minifier-plugin/tsconfig.json:4]: This should be the default.

File: webpack/module-minifier-plugin/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+    "lib": ["ES2017"],
+    "target": "ESNext",
```
Comment (iclanton) [inline webpack/module-minifier-plugin/tsconfig.json:6]: What is `es2017` missing?

File: webpack/module-minifier-plugin/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "./node_modules/@microsoft/rush-stack-compiler-3.5/includes/tsconfig-node.json",
+  "compilerOptions": {
+    "strictNullChecks": true,
+    "lib": ["ES2017"],
```
Comment (iclanton) [inline webpack/module-minifier-plugin/tsconfig.json:5]: This is the default.

**Comment by iclanton** [comment]
The `@ts-ignore`s make me uncomfortable. Can those be cleaned up?

File: common/reviews/api/module-minifier-plugin.api.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+export const IDENTIFIER_TRAILING_DIGITS: string;
+
+// @public
+export interface IExtendedModule extends webpack.compilation.Module, webpack.Module {
+    id: string | number | null;
+    identifier(): string;
+    readableIdentifier(requestShortener: webpack.compilation.RequestShortener): string;
+    resource?: string;
+    skipMinification?: boolean;
+}
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md:52]: If they tap the hook to customize the rehydration process, it is included in the input to said hook.

File: common/reviews/api/module-minifier-plugin.api.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+export const MODULE_WRAPPER_PREFIX = "__MINIFY_MODULE__(";
+
+// Warning: (ae-internal-missing-underscore) The name "MODULE_WRAPPER_SUFFIX" should be prefixed with an underscore because the declaration is marked as @internal
+//
+// @internal
+export const MODULE_WRAPPER_SUFFIX = ");";
+
+// @public
+export class ModuleMinifierPlugin {
+    constructor(options: IModuleMinifierPluginOptions);
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md:157]: The initial thought was that I would prefer to make minifier selection explicit. I would prefer for the default to be the WorkerPoolMinifier, but that requires `--experimental-worker` on node 10 and doesn't work at all on node 8. If I make a default I have to have it be `SynchronousMinifier` due to that limitation, and that's a suboptimal result on node 10. I suppose I could instead document that `--experimental-worker` is required on node 10, though.

File: common/reviews/api/module-minifier-plugin.api.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+// @internal
+export const MODULE_WRAPPER_PREFIX = "__MINIFY_MODULE__(";
+
+// Warning: (ae-internal-missing-underscore) The name "MODULE_WRAPPER_SUFFIX" should be prefixed with an underscore because the declaration is marked as @internal
+//
+// @internal
+export const MODULE_WRAPPER_SUFFIX = ");";
+
+// @public
+export class ModuleMinifierPlugin {
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md:156]: I had not realized that exists.

File: common/reviews/api/module-minifier-plugin.api.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    constructor(options: IModuleMinifierPluginOptions);
+    // (undocumented)
+    apply(compiler: webpack.Compiler): void;
+    // (undocumented)
+    readonly hooks: IModuleMinifierPluginHooks;
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md:161]: I think actually the correct course of action is to have the plugin add a hook to the compilation and then call it.

File: webpack/module-minifier-plugin/package.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    "type": "git",
+    "url": "https://github.com/Microsoft/rushstack/tree/master/webpack/module-minifier-plugin"
+  },
+  "scripts": {
+    "build": "gulp test --clean"
+  },
+  "peerDependencies": {
+    "webpack": "^4.31.0",
+    "webpack-sources": "~1.4.3",
+    "@types/webpack": "*"
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/package.json:18]: I cloned it from localization-plugin, though webpack-sources was new.

File: webpack/module-minifier-plugin/src/Constants.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+// Everything in this file is a `const` and TypeScript will infer more specific correct types than most developers would write.
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/src/Constants.ts:4]: Intellisense gives you the actual value without me having to write it twice.

File: webpack/module-minifier-plugin/src/GenerateLicenseFileForAsset.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+  let banner: string = '';
+
+  if (comments.size) {
+    // There are license comments in this chunk, so generate the companion file and inject a banner
+    const licenseSource: ConcatSource = new ConcatSource();
+    comments.forEach((comment) => {
+      licenseSource.add(comment);
+    });
+    const licenseFileName: string = `${assetName}.LICENSE.txt`;
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/src/GenerateLicenseFileForAsset.ts:45]: Huh. Excellent question.

File: webpack/module-minifier-plugin/tsconfig.json (hunk 3 above)
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/tsconfig.json:6]: `import()`, though I'm not currently using it.

File: build-tests/api-extractor-test-01/dist/api-extractor-test-01-beta.d.ts [hunk 12]
```diff
@@ -1,323 +0,0 @@
-/**
```
Comment (iclanton) [inline build-tests/api-extractor-test-01/dist/api-extractor-test-01-beta.d.ts:1]: I don't think these should be deleted.

File: common/reviews/api/module-minifier-plugin.api.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+export interface IModuleMinifierPluginHooks {
+    // (undocumented)
+    rehydrateAssets: AsyncSeriesWaterfallHook<IDehydratedAssets, webpack.compilation.Compilation>;
+}
+
+// @public
+export interface IModuleMinifierPluginOptions {
+    minifier: IModuleMinifier;
+    usePortableModules: boolean;
+}
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md]: Starting with no default minifier; will look at adding if it helps with adoption.

File: common/reviews/api/module-minifier-plugin.api.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply common/reviews/api/module-minifier-plugin.api.md:160]: We will look at making it optional in a later revision.

File: build-tests/api-extractor-test-01/dist/api-extractor-test-01-beta.d.ts (hunk 12 above)
Comment (dmichon-msft) [reply build-tests/api-extractor-test-01/dist/api-extractor-test-01-beta.d.ts]: Fixed.

File: webpack/module-minifier-plugin/src/GenerateLicenseFileForAsset.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (iclanton) [reply webpack/module-minifier-plugin/src/GenerateLicenseFileForAsset.ts:45]: Did you figure this out?

[the remark this answers is not in the document]
Comment (dmichon-msft) [reply webpack/module-minifier-plugin/src/GenerateLicenseFileForAsset.ts:45]: The final formatted name of the asset gets forwarded. So the [hash] macro expands to the hash of the original asset.

**Pull request merged**